### PR TITLE
Mix in chain + some immutability

### PR DESCRIPTION
**Pull request by koresar**
@ericelliott please review carefully.

This is not the final PR. One more to go to reach full immutability.

### CODE REVIEW

**Comment by koresar** [comment]
To simplify the review I will list changes:
 1. The `stamp.fixed` is not sharing anything with other stamps anymore. Plus unit tests for that.
 1. The `merge` and `mixin` functions of `mout` module as well as `mixinchain.js` function were replaced with enhanced (mostly copy&pasted) versions. Now we can control what's being copied/merged, and should it traverse prototype chain. Plus unit tests for that.
 1. The `Object.create` is reused (if present). Otherwise use the polyfil.
 1. The `methods` are now being copied from the given object **own** properties but not its prototype chain. Plus unit tests for that.
 1. `convertContructor` behaves exactly the same as in stampit 1.x regardless the changes inside the function. Plus unit tests for that.
 1. Unit tests were grouped (poorly).

I am ready to do any changes you wish.

File: mixer.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        loop(obj, copier, target);
+      }
+    }
+    return target;
+  }
+};
+
+module.exports.chain = true;
+module.exports.copyAll = null;
+module.exports.copyProp = function (val) {
```
Comment (ericelliott) [inline mixer.js]: Poorly named. This implies that this function makes a copy of something, when it's really testing whether or not it's a function.

File: mixer.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  return typeof val !== 'function';
+};
+module.exports.copyFunc = function (val) {
```
Comment (ericelliott) [inline mixer.js]: Also poorly named, and this is really just a duplicate of the previous function with reversed logic.

I suggest merging these into one function called `isFunction()`.

What's worse though is that none of these are used in this module at all. They should be defined in the module that uses them, instead.

File: mixer.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      obj = arguments[i];
+      if (obj) {
+        loop(obj, copier, target);
+      }
+    }
+    return target;
+  }
+};
+
+module.exports.chain = true;
```
Comment (ericelliott) [inline mixer.js]: The whole point of this branch is to make stamps immutable, but then you hang public mutable state on an exported object. This doesn't make any sense to me, and feels like a potential source of bugs and confusion.

File: stampit.js [hunk 4]
```diff
@@ -8,17 +8,23 @@
...
 'use strict';
 var forEach = require('mout/array/forEach');
-var mixIn = require('mout/object/mixIn');
-var merge = require('mout/object/merge');
 var map = require('mout/array/map');
 var forOwn = require('mout/object/forOwn');
-var mixInChain = require('./mixinchain.js');
 var slice = [].slice;
 
+var mixer = require('./mixer');
```
Comment (ericelliott) [inline stampit.js]: Why do I have to look at 3 different files to figure out the logic you're setting up in 3 lines of code? =)

File: merger.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+          },
+          target);
+      }
+    }
+
+    return target;
+  }
+};
+
+module.exports.chain = true;
```
Comment (ericelliott) [inline merger.js]: This is all duplicated in `mixer.js`.

File: mixer.js (hunk 3 above)
Comment (koresar) [inline mixer.js]: I didn't quite get what you mean. Mixer.js is not related to stampit.
Would you please rephrase that?

File: stampit.js (hunk 4 above)
Comment (koresar) [inline stampit.js]: Probably because the variable names are confusing. Any suggestions?

File: merger.js (hunk 5 above)
Comment (koresar) [inline merger.js]: Half of it, or mostly. But not all. deepClone and isObject are used here. What's the suggestion?

Comment (koresar) [inline merger.js]: Oh, by 'this' you mean `,chain`. Right?

Comment (ericelliott) [inline merger.js]: Yeah, I mean the shared module exports.

**Comment by koresar** [comment]
Ok, @ericelliott, check it out.

**Comment by ericelliott** [comment]
Cool, I'll be out of touch for a couple weeks. I'll look at this when I get back. Thanks!

**Comment by koresar** [comment]
Implemented stamp immutability. Additionally I removed the triple copy&paste. The `stamp.fixed` modification code was in the: `stamp()`, `compose()` and `methods()`/`state()`/`enclose()` functions. Not any more.

**Comment by koresar** [comment]
@ericelliott I believe it's time to move on. Would you like to review the code?

NB: the PR destination branch is 'v2_0' not the `master'.

File: stampit.js [hunk 6]
```diff
@@ -8,25 +8,30 @@
...
 
-create = function (o) {
-  if (arguments.length > 1) {
-    throw new Error('Object.create implementation only accepts the first parameter.');
-  }
-  function F() {}
+var mixer = require('./mixer');
+var mixIn = mixer.getMixin(); // Regular mixin function.
+var mixInFunctions = mixer.getMixin(isFunction); // mixin for functions only.
+var mixInChainFunctions = mixer.getMixin(isFunction, true); // mixin for functions including prototype chain.
```
Comment (ericelliott) [inline stampit.js]: This function signature is a [boolean trap](http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html). Possible solution: named options hash.

File: stampit.js [hunk 7]
```diff
@@ -8,25 +8,30 @@
...
+}
 
-create = function (o) {
-  if (arguments.length > 1) {
-    throw new Error('Object.create implementation only accepts the first parameter.');
-  }
-  function F() {}
+var mixer = require('./mixer');
+var mixIn = mixer.getMixin(); // Regular mixin function.
+var mixInFunctions = mixer.getMixin(isFunction); // mixin for functions only.
```
Comment (ericelliott) [inline stampit.js]: This is overly complicated. I'd prefer it if you just export all the various types of functions you need. Keep the DOT (Do One Thing) principle in mind when you write modules. You DOT and DRY don't have to conflict. Each of the separate modules can share dependencies for code reuse. The point is that we shouldn't expose the complexity of configuring mixer functions to the main stampit module. It should be able to simply require the functions it needs by name.

For more on how to keep module interfaces simple, watch the talk [Simplicity Matters by Rich Hickey](https://www.youtube.com/watch?v=rI8tNMsozo0).

File: dist/stampit.js [hunk 8]
```diff
@@ -710,9 +738,13 @@ isStamp = function isStamp(obj) {
  * @return {Function}             A composable stampit factory
  *                                (aka stamp).
  */
-convertConstructor = function convertConstructor(Constructor) {
-  return stampit().methods(Constructor.prototype).enclose(Constructor);
-};
+function convertConstructor(Constructor) {
+  var stamp = stampit();
+  mixInChainFunctions(stamp.fixed.methods, Constructor.prototype);
+  stamp.fixed.state = mergeChain(stamp.fixed.state, Constructor.prototype);
```
Comment (ericelliott) [inline dist/stampit.js]: This makes the previous line redundant, since calls will find everything on the instance and never look at the prototype chain.

File: dist/stampit.js [hunk 9]
```diff
@@ -710,9 +738,13 @@ isStamp = function isStamp(obj) {
  * @return {Function}             A composable stampit factory
  *                                (aka stamp).
  */
-convertConstructor = function convertConstructor(Constructor) {
-  return stampit().methods(Constructor.prototype).enclose(Constructor);
-};
+function convertConstructor(Constructor) {
```
Comment (ericelliott) [inline dist/stampit.js]: Couldn't this whole thing be rewritten with no change in instance behavior as:

```js
  return stampit().state(Constructor.prototype).enclose(Constructor);
```

?

File: test/stampit-specs.js [hunk 10]
```diff
@@ -17,63 +18,45 @@ test('.create()', function () {
...
 
 test('stampit().methods()', function () {
-  var obj = stampit().methods({
-    foo: function () {
-      return 'foo';
-    },
-    methodOverride: function () {
-      return false;
-    }
+  var obj = stampit({
```
Comment (ericelliott) [inline test/stampit-specs.js]: Why did you change this so it no longer uses `.methods()` for the initial setup? You realize now, you're not simply testing `.methods()`? You're also testing the parameter passing, which means you may get a failure during the `.methods()` test which is actually caused by the parameter functionality...

File: test/stampit-specs.js [hunk 11]
```diff
@@ -87,28 +70,36 @@ test('stampit().methods(a, b)', function () {
...
+  
   equal(obj.foo.bar, 'bar',
     'Should set default state.');
 });
 
 test('stampit().state()', function () {
-  var obj = stampit().state({
-    foo: {bar: 'bar'},
-    stateOverride: false
+  var obj = stampit(null, {
```
Comment (ericelliott) [inline test/stampit-specs.js]: This is also now testing both the parameter form and the `.state()` form. Why are you introducing these test ambiguities?

File: test/stampit-specs.js [hunk 12]
```diff
@@ -277,104 +345,58 @@ test('stampit.convertConstructor()', function () {
...
 
-  equal(u.bar(), 'bar',
-    'Should be able to add new methods with .compose()');
-
-  equal(u.baz, 'baz',
-    'Should be able to add new methods with .compose()');
-
-});
-
-test('stampit.compose() with inheritance', function () {
```
Comment (ericelliott) [inline test/stampit-specs.js]: Why did you delete this test?

File: stampit.js (hunk 6 above)
Comment (koresar) [inline stampit.js]: Thanks for the link. You are right.

File: stampit.js (hunk 7 above)
Comment (koresar) [inline stampit.js]: Got it.

File: dist/stampit.js (hunk 8 above)
Comment (koresar) [inline dist/stampit.js]: I'll check it.

File: dist/stampit.js (hunk 9 above)
Comment (koresar) [inline dist/stampit.js]: The `convertConstructor` doesn't work for old school classes which inherit other classes. Moreover, if the `Constructor` has some non prototype properties then they are not copied over either.

Basically, the two new comprehensive `'stampit.convertConstructor()'` tests fail.

File: test/stampit-specs.js (hunk 10 above)
Comment (koresar) [inline test/stampit-specs.js]: Sorry. I see your point.

File: test/stampit-specs.js (hunk 11 above)
Comment (koresar) [inline test/stampit-specs.js]: +1

File: test/stampit-specs.js (hunk 12 above)
Comment (koresar) [inline test/stampit-specs.js]: I didn't. I expanded it to test many-many other cases. See explanation above.

File: dist/stampit.js (hunk 9 above)
Comment (ericelliott) [inline dist/stampit.js]: Note that I replaced `.methods()` with `.state()`.

Does `.state()` follow the prototype chain or only work for own properties?

Comment (koresar) [inline dist/stampit.js]: Yeah. I noted. My comments above are related to your sample code.

Check it yourself :)

File: stampit.js (hunk 7 above)
Comment (koresar) [inline stampit.js]: Great talk!

File: dist/stampit.js (hunk 8 above)
Comment (koresar) [inline dist/stampit.js]: You are correct. Also, the 'converted' objects are not lightweight. Is that okay?

Comment (ericelliott) [inline dist/stampit.js]: What is the objective you're trying to accomplish by using a state stamp here?

Would this still be necessary if we stopped filtering `.methods()` to prevent state from being stored on the prototype?

Should we even be doing that?

File: stampit.js (hunk 7 above)
Comment (ericelliott) [inline stampit.js]: :+1:

File: dist/stampit.js (hunk 8 above)
Comment (koresar) [inline dist/stampit.js]: You are looking at the wrong commit. Please, check the combined changes. Here is the link to the one:
https://github.com/ericelliott/stampit/pull/48/files#diff-f547c18e2953252cb96744460b29a7cfR205

**Comment by koresar** [comment]
@ericelliott I have just pushed changes which address all your concerns.

**Comment by koresar** [comment]
Please see few more commits. Improved DOT and DRY of stampit.js file.

Created two more tests to improve DOT as suggested. They cover the bug #40.

**Comment by ericelliott** [comment]
Cool.. I'll check out the updated diff as soon as I can.

File: stampit.js [hunk 13]
```diff
@@ -189,16 +191,21 @@ isStamp = function isStamp(obj) {
  * @return {Function}             A composable stampit factory
  *                                (aka stamp).
  */
-convertConstructor = function convertConstructor(Constructor) {
-  return stampit().methods(Constructor.prototype).enclose(Constructor);
-};
+function convertConstructor(Constructor) {
+  var stamp = stampit();
+  mixer.mixInChainFunctions(stamp.fixed.methods, Constructor.prototype);
+  mixer.mixIn(stamp.fixed.state, Constructor);
```
Comment (ericelliott) [inline stampit.js]: What's the point of this? Very few Constructor functions have properties attached to them, and even if they did, it would be more appropriate to copy them to the stamp function itself than it would be to copy it as instance state.

File: stampit.js [hunk 14]
```diff
@@ -189,16 +191,21 @@ isStamp = function isStamp(obj) {
...
+  stamp.fixed.state = mixer.mergeChainObjects(stamp.fixed.state, Constructor.prototype);
```
Comment (ericelliott) [inline stampit.js]: Isn't it true that this will still override any method copied to `stamp.fixed.methods`? So this makes line 196 useless...

File: stampit.js (hunk 13 above)
Comment (koresar) [inline stampit.js]: If it converts classic constructors partially then we have to rename the function to `convertSomePropertiesOfTheConstructorButNotAll`. :)))

People usually have no idea if a constructor has some static props. For example `mongoose.Model`. It's got them all.

Sample code:
```js
function Constructor() {}
Constructor.clone = function (constructorObject) { ...
```

What's your proposal?

File: stampit.js (hunk 14 above)
Comment (koresar) [inline stampit.js]: My bad. It's a badly named function. `mergeChainObjects` should be probably named as `mergeChainNonFunctionProperties`.

It merges everything which is not functions.

File: stampit.js (hunk 13 above)
Comment (koresar) [inline stampit.js]: Oh, I see your point! Sorry being so stupid. :)
I take my words back. The line 197 should be eliminated I think.

Comment (koresar) [inline stampit.js]: Removed the line. See latest commit.

I am so sorry for the time taken.

**Comment by ericelliott** [comment]
There are lots of test failures in this branch. Can you double check that you're not using any ES5 features without polyfills, etc...?

Please feel free to use Mout or Lodash functions instead of ES5 features.

All tests are broken on:

* Internet Explorer 8
* iPhone 5

The last test is broken on Android 4

**Comment by ericelliott** [comment]
I think the Sauce Connect config is broken, but you can test manually using Sauce Labs.

**Comment by koresar** [comment]
I'll check it.

**Comment by koresar** [comment]
iPhone [passes](https://assets.saucelabs.com/jobs/bf326f8026d04cf98d869b6754bb566a/0003screenshot.png) fine.

Android also fine. The last test is added by test driver. Not stampit related.

IE8 issue "Object doesn't support property or method" can't fix myself. The `Object.create` shim is used but tests still fail. Digging in.

**Comment by koresar** [comment]
Ok, all fixed. The 'bind' function does not exist in IE8. That was my first ever IE8 developing experience.

I would need to join "IE8 haters" community now. Who's in charge?
